### PR TITLE
Spell log fixes

### PR DESCRIPTION
**Pull request by mityada**
SMSG_SPELLLOGEXECUTE now has the correct structure and data for all effects and targets.

SMSG_SPELLDISPELLOG and SMSG_DISPEL_FAILED now have the correct structures.

### CODE REVIEW

**Comment by Laizerox** [comment]
I will review this today. As for design I have own opinions how you could do it more clear.

I will also cross-check this with client because in TBC blizzard made a small shift for effects.

I will get back on PC.

**Comment by mityada** [comment]
With the packets corrected, client shows the following messages for dispel:
> Kobold Geomancer's Frost Armor is removed.
You cast Purge on Kobold Geomancer.

The order doesn't look natural. I think SMSG_SPELLDISPELLOG and SMSG_DISPEL_FAILED should be moved out of the effect handler to, say, Spell::SendDispelLog, which then will be called after Spell::SendLogExecute.

**Comment by Laizerox** [comment]
In all of the packets sniffed from retail spell log is always last.

**Comment by Phatcat** [comment]
Looking natural is not always the same as being blizzlike, would just like to throw that out there ^^

**Comment by mityada** [comment]
SMSG_SPELLINSTAKILLLOG had the wrong structure too. Fixed it.

I'd say too much wrong packets.

**Comment by Laizerox** [comment]
https://gist.github.com/Laizerox/d32f4d57e2f8304d047bcf2713903e32

Here's the whole structure for anyone intrested.

**Comment by Laizerox** [comment]
@mityada why don't we just use ByteBuffer and append to packet data in a loop? For example trinity does it that way and to preserve compatibility with TBC & WOTLK it would be quite mandatory. Because in those 2 client we're sending SpellLog for all summoned guids also. (For example 3x guardians)

Also the packets follow almost generic structure.

File: src/game/SpellEffects.cpp [hunk 1]
```diff
@@ -4894,7 +4940,14 @@ void Spell::EffectDurabilityDamage(SpellEffectIndex eff_idx)
         return;
 
     if (Item* item = ((Player*)unitTarget)->GetItemByPos(INVENTORY_SLOT_BAG_0, slot))
+    {
         ((Player*)unitTarget)->DurabilityPointsLoss(item, damage);
+
+        ExecuteLogInfo info(unitTarget->GetObjectGuid());
+        info.durabilityDamage.itemEntry = item->GetProto()->ItemId;
+        info.durabilityDamage.unk = -1;
```
Comment (Laizerox) [inline src/game/SpellEffects.cpp]: This is item slot

File: src/game/Spell.cpp [hunk 2]
```diff
@@ -3391,106 +3391,78 @@ void Spell::WriteSpellGoTargets(WorldPacket* data)
...
+
+    data << uint32(effectCount);
+
+    for (uint32 i = 0; i < MAX_EFFECT_INDEX; ++i)
+    {
+        if (m_executeLogInfo[i].empty())
+            continue;
+
+        data << uint32(m_spellInfo->Effect[i]);
+        data << uint32(m_executeLogInfo[i].size());
```
Comment (Laizerox) [inline src/game/Spell.cpp]: This is actually target counter

**Comment by mityada** [comment]
@Laizerox, do you mean using ByteBuffer instead of ExecuteLogInfo structure?

Anyway, why the current way is not compatible?
> Because in those 2 client we're sending SpellLog for all summoned guids also. (For example 3x guardians)

It's possible to have any number of ExecuteLogInfo per effect, so I don't see a problem here.

File: src/game/SpellEffects.cpp (hunk 1 above)
Comment (mityada) [inline src/game/SpellEffects.cpp]: How can you tell? That value is not actually used in the client.

**Comment by Laizerox** [comment]
@mityada yes there's no problem.

I had my own version based on trinitys now I mixed some of your ideas and will post a PR for TBC soon so we can see where we are at.

File: src/game/SpellEffects.cpp (hunk 1 above)
Comment (Laizerox) [inline src/game/SpellEffects.cpp]: It is. Take a closer to look to a structure here: https://gist.github.com/Laizerox/d32f4d57e2f8304d047bcf2713903e32#file-spelllogexecute-cpp-L195 and here https://gist.github.com/Laizerox/d32f4d57e2f8304d047bcf2713903e32#file-spelllogexecute-cpp-L196 and here https://gist.github.com/Laizerox/d32f4d57e2f8304d047bcf2713903e32#file-spelllogexecute-cpp-L200

Comment (mityada) [inline src/game/SpellEffects.cpp]: Of course I've seen that. I meant it's not used in any way that can tell it's an item slot.

**Comment by mityada** [comment]
I looked at the TrinityCore's code for this and I find it very similar. The only difference is that they have a separate structure for each case, while I use union.

**Comment by Laizerox** [comment]
@mityada I posted my solution written few weeks back and refactored with your idea here cmangos/mangos-tbc#126

**Comment by Laizerox** [comment]
@mityada could you please separate the SpellLogExecute and other packets into 2 different PR's?

**Pull request closed without merging**